File: src/main/java/work/lclpnet/ap2/base/ApActors.java
package work.lclpnet.ap2.base;

import net.minecraft.util.Util;
import work.lclpnet.ap2.api.actor.ActorProvider;
import work.lclpnet.ap2.api.actor.ActorRegistrar;
import work.lclpnet.ap2.impl.actor.GravityFieldActor;

import static work.lclpnet.ap2.api.actor.ActorFactory.withData;
import static work.lclpnet.ap2.base.ArcadeParty.identifier;
import static work.lclpnet.ap2.base.ArcadeParty.logger;

/**
 * An {@link ActorProvider} that provides all actor types from arcade-party-2.
 * Instances of this class exist outside the game-instance-scope.
 * An actor manager can receive this instance through a Fabric entrypoint.
 */
public class ApActors implements ActorProvider {

    @Override
    public void provideActors(ActorRegistrar registrar) {
        registrar.register(identifier("gravity_field"), withData(GravityFieldActor.Data.CODEC,
                Util.addPrefix("Parse GravityField data: ", logger::error),
                GravityFieldActor::new));
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/ApActors.kt
package work.lclpnet.ap2

import work.lclpnet.ap2.ApConstants.identifier
import work.lclpnet.ap2.api.actor.*
import work.lclpnet.ap2.impl.actor.GravityFieldActor

/**
 * An [ActorProvider] that provides all actor types from arcade-party-2.
 * Instances of this class exist outside the game-instance-scope.
 * An actor manager can receive this instance through a Fabric entrypoint.
 */
class ApActors : ActorProvider {

    override fun provideActors(registrar: ActorRegistrar) {
        registrar.register(GRAVITY_FIELD)
    }

    companion object {
        val GRAVITY_FIELD: ActorType<GravityFieldActor> = ActorType(
            identifier("gravity_field"),
            ActorFactory.withData(
                GravityFieldActor.Data.CODEC,
                { msg ->
                    ApConstants.logger.error("Parse GravityField data: $msg")
                }
            ) { init, data ->
                GravityFieldActor(init, data)
            }
        )
    }
}
